### PR TITLE
Make Option<TinyAsciiStr> to be same size as TinyAsciiStr

### LINKED ISSUE #2083: Make Option<TinyAsciiStr> be the same size as TinyAsciiStr
Previously, the `tinystr` crate used `NonZeroU32` and friends to optimize `Option<TinyStr4>`. We have not yet replicated this optimization after the rewrite to `TinyAsciiStr`.

@Manishearth had some suggestions on how to go about this.

### PR DESCRIPTION
**Pull request by pdogr**
Fixes #2083

### CODE REVIEW

File: utils/tinystr/src/ascii.rs [hunk 1]
```diff
@@ -110,30 +114,33 @@ impl<const N: usize> TinyAsciiStr<N> {
     #[must_use]
     pub fn len(&self) -> usize {
         if N <= 4 {
-            Aligned4::from_bytes(&self.bytes).len()
+            Aligned4::from_asciibytes(&self.bytes).len()
```
Comment (robertbastian) [inline utils/tinystr/src/ascii.rs:117]: Nit: `from_ascii_bytes`

File: utils/tinystr/src/ascii.rs [hunk 2]
```diff
@@ -110,30 +114,33 @@ impl<const N: usize> TinyAsciiStr<N> {
...
     pub const fn is_empty(&self) -> bool {
-        self.bytes[0] == 0
+        self.bytes[0] as u8 == AsciiByte::B0 as u8
     }
 
     #[inline]
     #[must_use]
     pub fn as_bytes(&self) -> &[u8] {
-        &self.bytes[0..self.len()]
+        unsafe { core::mem::transmute(&self.bytes[0..self.len()]) }
```
Comment (robertbastian) [inline utils/tinystr/src/ascii.rs:137]: Annotate each unsafe block with why it's safe

File: utils/tinystr/src/ascii.rs (hunk 1 above)
Comment (pdogr) [reply utils/tinystr/src/ascii.rs]: Done.

File: utils/tinystr/src/ascii.rs (hunk 2 above)
Comment (pdogr) [reply utils/tinystr/src/ascii.rs:138]: Done.

**Review by robertbastian — APPROVED** [review]
lgtm, but let's wait for @sffc's review as well

**Review by sffc — APPROVED** [review]

File: utils/tinystr/src/ascii.rs [hunk 3]
```diff
@@ -93,7 +94,10 @@ impl<const N: usize> TinyAsciiStr<N> {
             return Err(TinyStrError::ContainsNull);
         }
 
-        Ok(Self { bytes: out })
+        Ok(Self {
+            // SAFETY: `out` only contains ASCII bytes and has same size as `self.bytes`
+            bytes: unsafe { *(&out as *const [u8; N] as *const [AsciiByte; N]) },
+        })
```
[the remark this answers is not in the document]
Comment (pdogr) [reply utils/tinystr/src/ascii.rs]: Created `AsciiByte::to_ascii_byte_array` for this.

File: utils/tinystr/src/asciibyte.rs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    B124 = 124,
+    B125 = 125,
+    B126 = 126,
+    B127 = 127,
+}
+
+impl AsciiByte {
+    // Convert [u8; N] to [AsciiByte; N]
+    // SAFETY: AsciiByte is repr(u8) and same size as u8
+    pub const fn to_ascii_byte_array<const N: usize>(bytes: &[u8; N]) -> [AsciiByte; N] {
```
Comment (sffc) [inline utils/tinystr/src/asciibyte.rs:142]: Issue: The function should still be `unsafe`; it's just that the cast is centralized

Comment (sffc) [inline utils/tinystr/src/asciibyte.rs:142]: Nit: `#[inline]`

Comment (pdogr) [reply utils/tinystr/src/asciibyte.rs]: Done.

Comment (pdogr) [reply utils/tinystr/src/asciibyte.rs]: Done.

File: utils/tinystr/src/asciibyte.rs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    B123 = 123,
+    B124 = 124,
+    B125 = 125,
+    B126 = 126,
+    B127 = 127,
+}
+
+impl AsciiByte {
+    // Convert [u8; N] to [AsciiByte; N]
+    // SAFETY: AsciiByte is repr(u8) and same size as u8
```
Comment (robertbastian) [inline utils/tinystr/src/asciibyte.rs]: The safety invariant here should include that all bytes < 128

Comment (robertbastian) [reply utils/tinystr/src/asciibyte.rs]: whoops this was old code

**Review by robertbastian — APPROVED** [review]

**Pull request merged**